### DISCUSSION TITLE
Improve support of nested models

### DISCUSSION THREAD

**Discussion by lucianopaz** [pymc-devs/pymc3#5341] [Ideas]
At the moment, `pymc.Model` instances serve two purposes:

1. Book keeping of the different random variables that are created within their context.
2. Namespaces, where each random variable that is created within the context gets the model's name as a prefix to its own name.

The book keeping functionality enables nice features out of the box, like being able to draw the probabilistic graph of only a subset of random variables with graphviz:

```python
with pm.Model(name="m1") as m1:
    a = pm.Normal("a")
    with pm.Model(name="m2") as m2:
        b = pm.Normal("b")
        c = pm.HalfNormal("c")
        d = pm.Normal("d", b, c, shape=10)
    e = pm.Normal("e", a + d, shape=10)
pm.model_to_graphviz(m1)
pm.model_to_graphviz(m2)
```

renders the following graphs:

![imagen](https://user-images.githubusercontent.com/5230109/148986384-a887c373-fb9c-4a3f-8545-3349a703226e.png)
![imagen](https://user-images.githubusercontent.com/5230109/148986426-7b2a718a-63b6-4516-ba5f-aeae37ab5683.png)

The main problem is that the child model does not inherit its parents coords. Note that this snippet leads to an error because `m2` does not define the "dim" dimension:
```python
COORDS = {"dim": range(10)}
with pm.Model(name="m1", coords=COORDS) as m1:
    a = pm.Normal("a")
    with pm.Model(name="m2") as m2:
        b = pm.Normal("b")
        c = pm.HalfNormal("c")
        d = pm.Normal("d", b, c, dims="dim")
    e = pm.Normal("e", a + d, dims="dim")
```

One other problem is that the nested model's name completely supresses the parent model's name. Note that `d.name` is "m2_d" and not "m1_m2_d", which is more consistent in my opinion.

**I've opened this discussion to ask if there's any interest in having nested models inherit their parent's coordinates, and whether their name should also include their parent's name, if any.**

The first point is fairly straightforward to implement. We only need to copy the pattern that is used in the `free_RVs`, etc, attributes as done [here](https://github.com/pymc-devs/pymc/blob/main/pymc/model.py#L648-L665) and apply it also to the model's `_coords`, `_RV_dims` and `_dim_lengths` attributes.

The point about the model name should be even simpler to implement.

**Comment by ricardoV94** [comment]
Is there a case not to share dims by default?

**Reply by lucianopaz** [reply]
I don't think so. I think that the parent model should be made aware of any and all changes to the coords done by a nested model, and the nested models should be able to use the coords that were defined in the parent model. This parent/child sharing essentially means that sibling models would also know of the coords from their siblings.

**Reply by ricardoV94** [reply]
I agree, I think we can open an issue to fix the coords, and leave the discussion open for the name prefix.

**Comment by lucianopaz** [comment]
I just opened #5344 that should handle the nested model dims and coords. What should we do about the name prefix? I think the best thing to do would be to always add the parent model name as a prefix for the nested model's name.

**Comment by ferrine** [comment]
What do you think of / as scope separator?
`m1/m2/var`

**Reply by ricardoV94** [reply]
Sounds good. If there are no issues with InferenceData because of the slashes, I like it even more than the underscores

**Reply by lucianopaz** [reply]
I think that slash is a way better alternative to the underscore

**Comment by ferrine** [comment]
I was aware about limitations and had plans to improve it

**Comment by ferrine** [comment]
@ricardoV94 slash does not seem to break anything. PyMC+Xarray+Arviz
![image](https://user-images.githubusercontent.com/11705326/158784318-3c558502-5591-4878-81a2-92ad6c40320e.png)

**Reply by ricardoV94** [reply]
Perfect, wanna open a PR for this?

**Reply by ferrine** [reply]
YES
